### ISSUE TITLE
Another dataset but got 0&nan results

### ISSUE THREAD

**Issue by Rudder573** [blaz-r/SuperSimpleNet#7]
How amazing your project is! Hey, I try to run your project on other datasets like KolektorSDD2, and I put it in the 'mvtec' catalogue. But I got no results: `I-AUROC: 0.0 AP-det: nan seg-I-AUROC: 0.0 seg-AP-det: nan P-AUROC: 0.0 AUPRO: nan AP-loc: nan`. The dataset link is [https://github.com/2Obe/BSData](url). I want to know if  it is suitable for the SuperSimpleNet. Please let me know the fact at your convenience.

**Comment by ninechi143** [comment]
There can be many reasons which lead to this phenomenon, some of them are related to the hyperparameter settings and model computation flow, while the others are about the external bugs (which are irrelevant to Deep Learning algorithm, e.g., the format and expression of your annotation files are somehow incorrect).

If you are certain that the dataset and its format are complete,
I would like to say something happened in your model training, like, the model outputs, losses, or gradients ever exploded in one of parameter updates.

You can check the corresponding values in each iteration, i.e., in train.py, examine the values of losses when you feed each batch in train_loader.

**Comment by Rudder573** [comment]
Well, I received `UserWarning: No positive samples in targets, true positive value should be meaningless. Returning zero tensor in true positive score warnings.warn(*args, **kwargs)`. And the history result is like: 
wandb:      AP-det ▁▃▅▆▇▃▅▆▆▆█▃▄▆▃▇█▃█▆▇▅▃▅▅▅▅▄▅▆▅▄▇▆▆▅▅▅▄▄
wandb:       AUPRO ▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁
wandb:     I-AUROC ▁▃▄▅▆▆▃▃▄▄▅▂▇▂▁▄▅▂▅▁▇▁▆▇▃▃▃▄▃▄▃▄▄▄█▄▃▃▂▂
wandb:     P-AUROC ▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁
wandb:    avg_loss █▃▃▃▃▃▃▁▂▃▂▃▂▂▂▁▂▃▂▂▂▂▂▂▃▂▂▂▂▁▂▂▂▂▂▂▂▂▁▂
wandb:  batch_loss ▃▃▇▂▂▂▇▅▅▇▂▂▂▇▇▂▆▇▇▆▇▇▂▆▆▇▆▆▁▇▇█▁█▇▆▅▃▅▅
wandb:  seg-AP-det ▅█▄▁▆▃▃▂▅▁▃▃▂▃▁▄▄▁▁▁▂▃▂▃▂▃▂▂▁▄▂▂▂▃▃▃▂▂▂▂
wandb: seg-I-AUROC ▅▄▄▃█▁▁▆▄▄▂▂▁▆▄▁▂▆▃▂▁▄▃▁▃▃▃▂▃▂▂▃▃▂▂▃▂▃▃▂
Furthermore, the final result is `I-AUROC: 0.396806538105011 AP-det: 0.8442203402519226 seg-I-AUROC: 0.13888710737228394 seg-AP-det: 0.6895762085914612 P-AUROC: 0.0 AUPRO: 0 AP-loc: nan` 
I found the visualizing result as follows: 
![Image](https://github.com/user-attachments/assets/af30145a-16f8-44a6-84ad-f1f426265d23)
![Image](https://github.com/user-attachments/assets/002c811a-19ca-4804-ab1a-10cf10d648c5)
, which means however anomalous the sample is, the predicted mask is anomalous, and even the anomalous mask is not predicted. So I doubted if the dataset is suitable for this excellent program.

**Comment by blaz-r** [comment]
I will need more details to help, like which datasets and the results. Could you also please open a separate issue, since this looks unrelated to this one. Thank you.

**Comment by jin-123-456** [comment]
hank you for your reply. I've already created a new conversation.

**Comment by Rudder573** [comment]
> I believe this warning confirms what I said. Since there are no anomalous samples in the test set, the metric values are invalid.
> 
> Looking at the metric values, the image-level versions seem to have value, but the pixel ones don't. This would lead me to believe that something is wrong with how you load the dataset, especially the ground truth anomaly mask. Make sure the GT pixel-label in the test set is read as grayscale single-channel mask with anomalous pixel having value 1, and normal having value 0.

Well, I follow your advice, and it works. Thanks for your instructions. But I got bad results as well. 
```
wandb: Run history:
wandb:      AP-det ▁▅▆▅▃▄▅▅▅▆▇▃▂▃▅▅▂▇▇▅▇▇▃▅▆▂▄▅▄▇██▆▄▅▄▄▄▄▄
wandb:      AP-loc ▁▂▁▁▂▃▂█▁▁▁▁▂▁▁▁▁▁▁▁▂▂▁▂▁▁▁▁▁▂▂▁▁▁▂▂▁▁▂▂
wandb:       AUPRO ▁▇█▁▂▂▂▂▃▅▄▂▄▃▃▂▁▃▄▂▂▃▃▄▃▃▃▅▃▄▄▄▄▄▄▄▄▄▄▄
wandb:     I-AUROC ▁▃▄▅▅▂▄▃▄▄▅▇▁▂▄▄▅▂▇▄▆▆▁▃▃▃▄▂▃▅▄▆█▆▄▃▃▃▂▃
wandb:     P-AUROC ▃▃▃▅▄▄█▁▃▂▂▁▁▃▂▂▄▃▅▂▄▂▃▂▃▂▂▄▂▃▂▂▄▂▃▃▃▃▄▄
wandb:    avg_loss █▅▄▃▂▃▃▃▃▃▃▂▂▃▃▂▃▂▂▃▄▃▃▃▂▃▂▃▃▃▃▂▂▂▂▂▂▂▁▂
wandb:  batch_loss ▃▃▆█▆▂▂▆█▆█▆▂▁▄▁▆▅▁▂▆▆▂▇▆▂▂▁▆▁▆▇▁▂▇▅▅▅▅▇
wandb:  seg-AP-det ▅▆▄█▄▂▁▄▃▂▅▁▂▃▃▃▂▁▃▄▁▁▃▂▃▃▂▂▃▃▁▄▃▂▂▃▃▃▂▂
wandb: seg-I-AUROC ▄▃▄██▄▃▂▂▁▆▄▁▁▃▂▁▃▃▃▁▁▂▃▂▂▁▂▃▁▃▃▂▂▂▂▂▂▂▂
wandb: 
wandb: Run summary:
wandb:      AP-det 0.84422
wandb:      AP-loc 0.00231
wandb:       AUPRO 0.05912
wandb:     I-AUROC 0.39681
wandb:     P-AUROC 0.43062
wandb:    avg_loss 1.00132
wandb:  batch_loss 0.26962
wandb:  seg-AP-det 0.68958
wandb: seg-I-AUROC 0.13889
```
I set `config["perlin_thr"] = 0.7`, is that influence the performance? I found the visualizing result as follows:

![Image](https://github.com/user-attachments/assets/2c55c03a-424a-404b-a7cb-8b94edda18d6)

![Image](https://github.com/user-attachments/assets/8d17d6d8-49b7-4724-8727-848da605c55e)
The predicted masks have no difference between the good and anomalous samples.

**Comment by blaz-r** [comment]
I suggest you use the same setup as for `mvtec` for other unsupervised datasets. This means that you set `perlin_thr` to 0.2 initially. You could then try to increase it, but 0.2 is the most general option.

**Comment by Rudder573** [comment]
Thanks for your detailed explanation. It is true that the dataset is the BSData dataset. The normal samples only contain the top two images, and abnormal ones consist of both types. Maybe I delete the below types of images? Or set the zoom level to (512, 512) regarding to the orginal picture is (1130, 460) (1540, 645) (2724, 850)?

**Comment by Rudder573** [comment]
Thank you for your continuous support and for patiently answering all my questions. Your help means a lot to me! Well, I checked the dataset again. I found the good samples only contain Type1, while anomalous samples contain Type1 & 2. 
Type1_normal
![Image](https://github.com/user-attachments/assets/2961c25f-6a05-402d-915a-e12bf8af1737)
Type1_abnormal
![Image](https://github.com/user-attachments/assets/5249c566-1000-4ad4-bb12-28f6c2dfe06c)
Type2_abnormal
![Image](https://github.com/user-attachments/assets/a74380b6-12fa-48fb-ad93-40c3d1ae61e7)
When I deleted all type 2 pictures, I got nice results:
```
wandb: Run summary:
wandb:      AP-det 0.96888
wandb:      AP-loc 0.29414
wandb:      AUPRO 0.84696
wandb:      I-AUROC 0.86497
wandb:      P-AUROC 0.98591
wandb:      avg_loss 0.00037
wandb:      batch_loss 0
wandb:      seg-AP-det 0.98838
wandb:      seg-I-AUROC 0.94894
```
Visualizing results are as follows:
![Image](https://github.com/user-attachments/assets/9267e647-4d92-4e27-ac1a-28535ed51dad)
However, there're some unpredicted masks:
![Image](https://github.com/user-attachments/assets/b5ea96ba-3f12-4c74-972e-a3373c157e66)
![Image](https://github.com/user-attachments/assets/c616a267-89ca-4094-b84a-11dba63fa677)
Well, you can see the anomaly map received the abnormal pixels indeed. I know this dataset may not be suitable for unsupervised anomaly detection, but my spirit originated from this article: [A Double Self-Supervised Model for Pitting Detection on Ball Screws](https://ieeexplore.ieee.org/abstract/document/10479204). It is for my master thesis, and I already done the supervised method for accurate semantic segmentation. Furthermore, I want to solve the problem 'overdependence on data annotations' , so I choose this method. I really appreciate all the time and effort you've put into answering my questions so thoroughly. Your patience and kindness truly make this project even more inspiring. Thank you!

**Comment by blaz-r** [comment]
About the unpredicted masks. I'd mostly focus on anomaly map, since the predicted mask is just anomal map thresholded with 0.5. Since all the metrics are threshold agnostic, this 0.5 mioght not be best for visualisation, it was just arbitrarly chosen. From these results it looks like the model is struggling a bit, since you can see that seg-I-AUROC (where anomaly detection is based on max value in anomaly map) works better than the I-AUROC (anomaly score is from classification head). 
This can happen for some datasets, so you could keep the anomaly score from classification head, or if you want better results use the max value of anomaly map (seg-I-AUROC) which is mroe in line with what they show in the paper you sent.

Metric curves can be unstable since the model is trained with random noise. I think that's fine for your run. What batch size are you using? Larger batch sizes usually lead to more stable trainign and usually quite better results.

**Comment by Rudder573** [comment]
Well, where can I revise the anomal map threshold? Batchsize is 4 because of only 1 gpu: rtx 3080 Ti. What I said is epochs, which are increased  from 300 to 500, but the model tends to overfit.

**Comment by blaz-r** [comment]
Larger batch size will definitely help, longer epochs probably not. The threshold needs to be manually set inside the trst function.

**Comment by Rudder573** [comment]
Thank you very much for your kind guidance — I truly appreciate your help. With your support, I was able to achieve quite promising results.

I would like to further improve the accuracy of defect localization. Specifically, I’m hoping to make the predicted results more closely match the actual defect contours. Because there are some micro defects which cannot be detected, this may lead to low image-level scores. The results are as follows:
The non-detected pic:
<img width="1138" height="245" alt="Image" src="https://github.com/user-attachments/assets/2a26cf29-f750-435a-89ac-40d01a4a5fc2" />
The contour is not close to the true mask:
<img width="1138" height="245" alt="Image" src="https://github.com/user-attachments/assets/bcc44484-8e01-4122-8d30-52181066ceb7" />

Could you kindly advise on what techniques or post-processing steps I might consider to achieve this?

Looking forward to your insights again, and thank you once more for your excellent work on the project.

**Comment by blaz-r** [comment]
I'm closing this due to inactivity. Feel free to reopen or create a new issue if something new arises.
